File: 0001-0099/0007. Reverse Integer/0007. Reverse Integer.py
class Solution:
    def reverse(self, x: int) -> int:
        max_int = 2 ** 31 - 1
        min_int = -(2 ** 31)
        res = 0
        while x != 0:
            tail = x % 10 if x > 0 else x % -10
            if (
                x > 0
                and (
                    max_int // 10 > res
                    or (max_int // 10 == res and max_int % 10 > tail)
                )
            ) or (
                x < 0
                and (
                    int(min_int / 10) < res
                    or (int(min_int / 10) == res and min_int % -10 < tail)
                )
            ):
                res = res * 10 + tail
                x = int(x / 10)
            else:
                return 0
        return res

File: 0001-0099/0007. Reverse Integer/0007. Reverse Integer.js
const reverse = function(x) {
    let result = 0;
    while (x !== 0) {
        const tail = x % 10;
        const newResult = (Math.imul(result,10) + tail) | 0
        if (Math.trunc(newResult / 10) !== result) return 0  // 这里不需要减去tail 因为除10自然去掉个位
        result = newResult;
        x = Math.trunc(x / 10)
    }
    return result;
};
//位运算会强制使用32位整数
// 由于 |0这种做法 对于乘法不起作用 需要Math.mul
// const a = 210 | 0; // 强制使用32bit整数来模拟此题目的情况
// https://stackoverflow.com/questions/19843652/32-bit-signed-integer-math-in-javascript

